symplyphysics/core/experimental/coordinate_systems/__init__.py

```diff
@@ -0,0 +1,199 @@
+"""
+Coordinate systems
+==================
+
+In this module the base class for coordinate systems is defined along with several
+most commonly used systems: Cartesian, cylindrical, and spherical systems.
+
+If you want to add a custom coordinate system, subclass ``CurvilinearCoordinateSystem``
+and define the scalar names and the Lamé coefficients related to it.
+
+**Limitations:**
+
+1. The systems share the same point of origin and are defined in relation to the same
+   axes. The rotation and translation of coordinate systems are not supported.
+
+2. The systems are orthogonal and right-handed.
+
+3. The space is 3-dimensional.
+"""
+
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+
+from sympy import Expr, Symbol, symbols, S, sin
+
+
+class BaseCoordinateSystem(metaclass=ABCMeta):
+    """
+    Abstract base class for all coordinate systems.
+    """
+
+    _display_name: str
+    _base_scalars: tuple[Symbol, Symbol, Symbol]
+    _lame_coefficients: tuple[Expr, Expr, Expr]
+
+    @classmethod
+    @abstractmethod
+    def scalar_names(cls) -> tuple[str, str, str]:
+        """
+        Names of the base scalars.
+        """
+
+    @abstractmethod
+    def __init__(self, display_name: str) -> None:
+        pass
+
+    @property
+    def display_name(self) -> str:
+        """
+        Display name of the coordinate system.
+        """
+
+        return self._display_name
+
+    @property
+    def base_scalars(self) -> tuple[Expr, Expr, Expr]:
+        """
+        A 3-tuple of symbols for base scalars.
+        """
+
+        return self._base_scalars
+
+    @property
+    def lame_coefficients(self) -> tuple[Expr, Expr, Expr]:
+        """
+        A 3-tuple of Lamé coefficients of the coordinate system.
+        """
+
+        return self._lame_coefficients
+
+    @property
+    def jacobian_determinant(self) -> Expr:
+        """
+        The determinant of the Jacobian matrix of transformation. Equivalently,
+        the product of all the Lamé coefficients.
+        """
+
+        lame_ = self.lame_coefficients
+        return lame_[0] * lame_[1] * lame_[2]
+
+    def __str__(self) -> str:
+        return self.display_name
+
+    def __eq__(self, other: object) -> bool:
+        return (
+            type(other) is type(self)
+            and self.lame_coefficients == other.lame_coefficients  # type: ignore[attr-defined]
+        )
+
+    def __getitem__(self, name: str) -> Expr:
+        names = self.scalar_names()
+        try:
+            index = names.index(name)
+        except ValueError as e:
+            raise IndexError(f"Expect a scalar name from {names}, got {name!r}.") from e
+
+        return self.base_scalars[index]
+
+
+def generate_scalar_names(system: BaseCoordinateSystem) -> tuple[str, str, str]:
+    """
+    Subscript the scalar names with the given suffix.
+    """
+
+    suffix = system.display_name
+    names = [f"{name}_{suffix}" for name in system.scalar_names()]
```

Comment:
I guess `x_C` makes more sense from a mathematical standpoint. Since `C` only gives additional information, it's reasonable to place it in the subscript.